lib/active_merchant/billing/gateways/trans_first_transaction_express.rb

```diff
@@ -0,0 +1,597 @@
+            add_amount(doc, amount)
+          end
+        else
+          request = build_xml_transaction_request do |doc|
+            add_transaction_code(doc, :authorize)
+            add_amount(doc, amount)
+            add_wallet_id(doc, payment_method)
+          end
+        end
+
+        commit(request)
+      end
+
+      def capture(amount, authorization, options={})
+        request = build_xml_transaction_request do |doc|
+          add_transaction_code(doc, :capture)
+          add_amount(doc, amount)
+          add_original_transaction_data(doc, authorization)
+        end
+
+        commit(request)
+      end
+
+      # Transaction Express has three types of possible void
+      # - void_authorization
+      # - void_purchase (alias: void_capture)
+      # - void_refund (alias: void_credit)
+      def void(authorization, options={})
+        void_type = (options[:void_type] || :void_purchase)
+
+        request = build_xml_transaction_request do |doc|
+          add_transaction_code(doc, void_type)
+          add_original_transaction_data(doc, authorization)
+        end
+
+        commit(request)
+      end
+
+      def refund(amount, authorization, options={})
+        request = build_xml_transaction_request do |doc|
+          add_transaction_code(doc, :refund)
+          add_amount(doc, amount)
+          add_original_transaction_data(doc, authorization)
+          add_order_number(doc, options)
+        end
+
+        commit(request)
+      end
+
+      def credit(amount, payment_method, options={})
+        request = build_xml_transaction_request do |doc|
+          add_transaction_code(doc, :blind_credit)
+          add_pan(doc, payment_method)
+          add_amount(doc, amount)
+        end
+
+        commit(request)
+      end
+
+      def verify(credit_card, options={})
+        request = build_xml_transaction_request do |doc|
+          add_transaction_code(doc, :verify)
+          add_payment_method(doc, credit_card)
+          add_contact(doc, credit_card.name, options)
+        end
+
+        commit(request)
+      end
+
+      def store(payment_method, options={})
+        store_customer_request = build_xml_payment_storage_request do |doc|
+          store_customer_details(doc, payment_method.name, options)
+        end
+
+        MultiResponse.run do |r|
+          r.process { commit(store_customer_request) }
+          return r unless r.success? && r.params["custId"]
+          customer_id = r.params["custId"]
+
+          store_payment_method_request = build_xml_payment_storage_request do |doc|
+            doc.tag! "v1:cust" do
+              add_customer_id(doc, customer_id)
+              doc.tag! "v1:pmt" do
+                doc.tag! "v1:type", 0 # add
+                add_payment_method(doc, payment_method)
+              end
+            end
+          end
+
+          r.process { commit(store_payment_method_request) }
+        end
+      end
+
+      def update(token, payment_method, options={})
```

Comment:
@rwdaigle some gateways do actually support it as part of the public AM API; that said, I think it's hard to support consistently so I'm :+1: on not making it part of the standard suite of AM methods we implement.